### PR TITLE
RABSW-1007: Check for unique fields

### PR DESCRIPTION
**Pull request by matthew-richerson**
This commit adds a new option for a rule definition, "uniqueWithin". This
option allows rules to specify that a key/value pair needs to be unique across
multiple #DWs. An example is the "name" field in a jobdw directive. Multiple
jobdws should not use the same value for name. However, The same value for name
can be used between a jobdw and a create_persistent. "uniqueBetween" allows for
this by specifying a key. Values will only be checked for uniqueness for rule
definitions using the same "uniqueBetween" key.

Signed-off-by: Matt Richerson <mattr@cray.com>

### CODE REVIEW

**Comment by roehrich-hpe** [comment]
"However, The same value for name
can be used between a jobdw and a create_persistent."

Why is that?  Are these #DW's in the same script?

File: utils/dwdparse/dwdparse_test.go [hunk 1]
```diff
@@ -238,174 +241,187 @@ const (
...
+	{[]string{"#DW jobdw type=xfs    capacity=100GB name=prettyGoodName  "}, allow, validDWOrAllowUnknownCommand},
+	{[]string{"#DW jobdw type=lustre capacity=100GB name=prettyGoodName  "}, allow, validDWOrAllowUnknownCommand},
+	{[]string{"#DW jobdw type=raw    capacity=100TB name=prettyGoodName  "}, allow, validDWOrAllowUnknownCommand},
+	{[]string{"#DW jobdw type=xfs    capacity=100TB name=prettyGoodName  "}, allow, validDWOrAllowUnknownCommand},
+	{[]string{"#DW jobdw type=lustre capacity=100TB name=prettyGoodName  "}, allow, validDWOrAllowUnknownCommand},
+	{[]string{"#DW jobdw type=raw    capacity=100GB name=prettierGoodName"}, allow, validDWOrAllowUnknownCommand},
+	{[]string{"#DW jobdw type=xfs    capacity=100GB name=prettierGoodName"}, allow, validDWOrAllowUnknownCommand},
+	{[]string{"#DW jobdw type=lustre capacity=100GB name=prettierGoodName"}, allow, validDWOrAllowUnknownCommand},
+
+	{[]string{"#DW jobdw type=raw    capacity=100GB name=uniqueName1 ", "#DW jobdw type=raw    capacity=100GB name=uniqueName2 "}, allow, validDWOrAllowUnknownCommand},
```
Comment (roehrich-hpe) [inline utils/dwdparse/dwdparse_test.go:270]: Can you split these lines so they stand out as multi-DW to the eye?  Will go fmt allow that?

**Review by roehrich-hpe — APPROVED** [review]

**Comment by matthew-richerson** [comment]
> 
Yeah a create_persistent and a jobdw can have the same name even if they're in the same job script. Any reference to the jobdw will be JOBDW_[name] and and reference to the persistent instance will be "PERSISTENTDW_[name]", so there won't be any conflicts.

**Comment by roehrich-hpe** [comment]
> Yeah a create_persistent and a jobdw can have the same name even if they're in the same job script. Any reference to the jobdw will be JOBDW_[name] and and reference to the persistent instance will be "PERSISTENTDW_[name]", so there won't be any conflicts.

That feels like it would be confusing.  I think "unique" should mean unique across the script.

**Comment by matthew-richerson** [comment]
> > Yeah a create_persistent and a jobdw can have the same name even if they're in the same job script. Any reference to the jobdw will be JOBDW_[name] and and reference to the persistent instance will be "PERSISTENTDW_[name]", so there won't be any conflicts.
> 
> That feels like it would be confusing. I think "unique" should mean unique across the script.

That's why it's "UniqueWithin". It's unique within a certain namespace. Either the jobdw or create_persistent namespace

**Pull request merged**